### ISSUE TITLE
How to handle mutable arguments?

### ISSUE THREAD

**Issue by fjarri** [getditto/safer_ffi#95]
I want to add error reporting in my API along the lines of
```rust
#[derive_ReprC]
#[repr(C)]
pub struct Error {
    error_code: u32,
    error_str: char_p::Box,
}
```

```c
Error error;
api_func(&error);
printf("%d: %s\n", error.error_code, error.error_str);
free_error(&error); // to free the memory allocated for `error_str`
```
(just an example, `api_func()` returns a result value too)

I tried various approaches on the Rust side, but I cannot make it work; at best it compiles, but in runtime I get a `malloc: *** error for object <...>: pointer being freed was not allocated` error. For example, for this code:
```rust
#[ffi_export]
fn api_func(error: &mut Error) {
    x.error_code = 0;
    x.error_str = "error".to_string().try_into().unwrap();
}
```
or for the example in #30. What is the correct way of doing it in `safer_ffi`? It seems that my problem is that I cannot figure out how to tell the compiler not to drop the newly created string.

(Ideally I'd want the reference to be nullable, so that one could call `api_func(NULL)` in C, and the error reporting would be skipped, but I'll probably be able to deduce it)

**Comment by danielhenrymantilla** [comment]
I'm going to close the issue since I consider that the usage of the optional `Out` type solves your use case. But don't let the closed issue deter you from asking further questions or clarifications! 🙂 In other words, should you encounter any issue when following my indications, you are free to keep asking about it, even if the issue appears to be closed

**Comment by fjarri** [comment]
Thanks for the detailed response, and sorry it took me so long to come up with a feedback. This was indeed very helpful — I did not notice `Out` in the documentation. The only problem I'm still having is that the compiler is not smart enough to figure out that I'm not reusing the `error` value if `.write()` is called in a nested method call. 

To illustrate, I have a bunch of methods that may return `Option` or `Result`, and I want to make it easy for the user to forward their outcome into the FFI error. Ideally, I would want to write something like
```rust
#[ffi_export]
fn api_func(error: Option<Out<'_, Error>>) -> Option<usize> {
    let x = returns_option().report_error("returns_option() failed", error)?;
    let y = returns_result().report_error("returns_result() failed", error)?;
    Some(x + y)
}
```
and `report_result()` is something like
```rust
impl<T, E: Display> Reportable<T> for Result<T, E> {
    fn report_error<'a>(self, message: &'a str, error: Option<Out<'a, Error>>) -> Option<T> {
        match self {
            Ok(value) => Some(value),
            Err(err) => {
                write_error(format!("{}: {}", message, err), error);
                None
            }
        }
    }
}
```
where `write_error()` consumes `error`. Similar implementation for `Option`.

The way `report_error()` is implemented, `error` is consumed on an erroneous variant of `Option` or `Result`, so theoretically the compiler should be able to deduce that the usage in `api_func()` is correct. And indeed it can deduce that if I inline `report_error()`. But without the inlining, alas, it complains about `error` being used after it's consumed in the first call. 

I see two ways out of this, both are not completely satisfactory.

Explicit: make `report_error(...) -> Option<(T, Option<Out<'a, Error>)>`. So `api_func()` would look like 
```rust
#[ffi_export]
fn api_func(error: Option<Out<'_, Error>>) -> Option<usize> {
    let (x, rust_error) = returns_option().report_error("returns_option() failed", error)?;
    let (y, _rust_error) = returns_result().report_error("returns_result() failed", error)?;
    Some(x + y)
}
```
This adds visual noise to the users of `report_error()`.

Macro: inline `repor_error()` every time. So `api_func()` would look like 
```rust
#[ffi_export]
fn api_func(error: Option<Out<'_, Error>>) -> Option<usize> {
    let x = report_option!(returns_option(), "returns_option() failed", error);
    let y = report_result!(returns_result(), "returns_result() failed", error);
    Some(x + y)
}
```
I consider macros the absolute last resort because they're hard to maintain and debug. And it makes the code too vague.

Any ideas?

**Comment by fjarri** [comment]
Thanks for the explanation, this makes sense. There's one problem though, a declaration like
```
  #[ffi_export]
  fn api_func(mut error: Option<Out<'_, Error>>) -> Option<usize> {
```
does not compile - I'm getting `error` underlined and `no rules expected this token in macro call` error. I'm using `safer-ffi=0.9`, is that fixed in the trunk?

**Comment by danielhenrymantilla** [comment]
The `mut error` binding got fixed recently: supported as of `0.0.10` (#111)

**Comment by danielhenrymantilla** [comment]
@fjarri I've moved the discussion and provided more info over the `Discussions/` are of this repo: #113
